controller/volume_controller_test.go

```diff
@@ -820,6 +821,87 @@ func (s *TestSuite) TestVolumeLifeCycle(c *C) {
 	}
 	tc.expectReplicas = expectRs
 	testCases["volume attaching - start replicas - node failed"] = tc
+
+	// Disable revision counter
+	tc = generateVolumeTestCaseTemplate()
+	tc.volume.Spec.NodeID = TestNode1
+	tc.volume.Status.OwnerID = TestNode1
+	tc.volume.Status.CurrentNodeID = TestNode1
+
+	// Since default node 2 is scheduling disabled,
+	// enable node soft anti-affinity for 2 replicas.
+	tc.replicaNodeSoftAntiAffinity = "true"
+	tc.volume.Spec.RevisionCounterDisabled = true
+	tc.copyCurrentToExpect()
+	tc.engines = nil
+	tc.replicas = nil
+
+	tc.expectVolume.Status.State = types.VolumeStateAttaching
+	tc.expectVolume.Status.CurrentImage = tc.volume.Spec.EngineImage
+	tc.expectVolume.Status.CurrentNodeID = tc.volume.Spec.NodeID
+	expectEs := map[string]*longhorn.Engine{}
+	for _, e := range tc.expectEngines {
+		e.Spec.RevisionCounterDisabled = true
+		expectEs[e.Name] = e
+	}
+	tc.expectEngines = expectEs
+	expectRs = map[string]*longhorn.Replica{}
+	for _, r := range tc.expectReplicas {
+		r.Spec.DesireState = types.InstanceStateRunning
+		r.Spec.RevisionCounterDisabled = true
+		expectRs[r.Name] = r
+	}
+	tc.expectReplicas = expectRs
+
+	testCases["volume revision counter disabled - engine and replica revision counter disabled"] = tc
+
+	// Salvage Requested
+	tc = generateVolumeTestCaseTemplate()
+	tc.volume.Spec.NodeID = TestNode1
+	tc.volume.Status.State = types.VolumeStateDetached
```

Comment:
`tc.volume.Status.State = types.VolumeStateAttached`

Reply:
Hi Shuo, I think it need to be detached [code is here](https://github.com/longhorn/longhorn-manager/blob/e4f9956d28fde33027eb3dd9b69f3456cea55c17/controller/volume_controller.go#L884), since this is one sync call. In the real case, it can be attached, and need to become detached first, then be salvaged. Please correct me if i am wrong. Thanks.

Reply:
In the init state, the salvaged hasn't been executed. And how can a `detached` volume suddenly becomes `detaching`.

Reply:
I think you are right, probably there is something wrong with my PR, let me check it again. Thanks.

Reply:
@shuo-wu , I have updated my PR, also fixed [this part](https://github.com/longhorn/longhorn-manager/blob/d94f2f28d8b76041650b0d979e88825f7b240368/controller/volume_controller.go#L953-L959), please take a look and let me know. Thanks.